core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -194,189 +216,149 @@ class TransactionCoordinator(brokerId: Int,
   }
 
-  def handleTxnEmigration(transactionStateTopicPartitionId: Int) {
-      txnManager.removeTransactionsForPartition(transactionStateTopicPartitionId)
-      txnMarkerChannelManager.removeStateForPartition(transactionStateTopicPartitionId)
+  def handleTxnEmigration(txnTopicPartitionId: Int) {
+      txnManager.removeTransactionsForTxnTopicPartition(txnTopicPartitionId)
+      txnMarkerChannelManager.removeMarkersForTxnTopicPartition(txnTopicPartitionId)
   }
 
   def handleEndTransaction(transactionalId: String,
                            pid: Long,
                            epoch: Short,
-                           command: TransactionResult,
+                           txnMarkerResult: TransactionResult,
                            responseCallback: EndTxnCallback): Unit = {
-    val errors = validateTransactionalId(transactionalId)
-    if (errors != Errors.NONE)
-      responseCallback(errors)
-    else
-      txnManager.getTransactionState(transactionalId) match {
+    val error = validateTransactionalId(transactionalId)
+    if (error != Errors.NONE)
+      responseCallback(error)
+    else {
+      val result: Either[Errors, (Int, TransactionMetadataTransition)] = txnManager.getTransactionState(transactionalId) match {
         case None =>
-          responseCallback(Errors.INVALID_PID_MAPPING)
-        case Some(metadata) =>
-          metadata synchronized {
-            if (metadata.pid != pid)
-              responseCallback(Errors.INVALID_PID_MAPPING)
-            else if (metadata.producerEpoch != epoch)
-              responseCallback(Errors.INVALID_PRODUCER_EPOCH)
-            else metadata.state match {
+          Left(Errors.INVALID_PID_MAPPING)
+
+        case Some(epochAndTxnMetadata) =>
+          val txnMetadata = epochAndTxnMetadata.transactionMetadata
+          val coordinatorEpoch = epochAndTxnMetadata.coordinatorEpoch
+
+          txnMetadata synchronized {
+            if (txnMetadata.producerId != pid)
+              Left(Errors.INVALID_PID_MAPPING)
+            else if (txnMetadata.producerEpoch != epoch)
+              Left(Errors.INVALID_PRODUCER_EPOCH)
+            else if (txnMetadata.pendingTransitionInProgress)
+              Left(Errors.CONCURRENT_TRANSACTIONS)
+            else txnMetadata.state match {
               case Ongoing =>
-                commitOrAbort(transactionalId, pid, epoch, command, responseCallback, metadata)
+                val nextState = if (txnMarkerResult == TransactionResult.COMMIT)
+                  PrepareCommit
+                else
+                  PrepareAbort
+                Right(coordinatorEpoch, txnMetadata.prepareAbortOrCommit(nextState, time.milliseconds()))
               case CompleteCommit =>
-                if (command == TransactionResult.COMMIT)
-                  responseCallback(Errors.NONE)
+                if (txnMarkerResult == TransactionResult.COMMIT)
+                  // this is an optimization: return OK immediately
```

Comment:
nit: I think this is more than an optimization: it is necessary for correctness because there is no guarantee that the client will receive the result of the EndTxnRequest.